synapse/handlers/jwt.py

```diff
@@ -0,0 +1,118 @@
+
+        Args:
+            login_submission: the whole of the login submission
+                (including 'type' and other relevant fields)
+
+        Returns:
+            The user ID that is logging in.
+
+        Raises:
+            LoginError if there was an authentication problem.
+        """
+        token = login_submission.get("token", None)
+        if token is None:
+            raise LoginError(
+                403, "Token field for JWT is missing", errcode=Codes.FORBIDDEN
+            )
+
+        jwt = JsonWebToken([self.jwt_algorithm])
+        claim_options = {}
+        if self.jwt_issuer is not None:
+            claim_options["iss"] = {"value": self.jwt_issuer, "essential": True}
+        if self.jwt_audiences is not None:
+            claim_options["aud"] = {"values": self.jwt_audiences, "essential": True}
+
+        try:
+            claims = jwt.decode(
+                token,
+                key=self.jwt_secret,
+                claims_cls=JWTClaims,
+                claims_options=claim_options,
+            )
+        except BadSignatureError:
+            # We handle this case separately to provide a better error message
+            raise LoginError(
+                403,
+                "JWT validation failed: Signature verification failed",
+                errcode=Codes.FORBIDDEN,
+            )
+        except JoseError as e:
+            # A JWT error occurred, return some info back to the client.
+            raise LoginError(
+                403,
+                "JWT validation failed: %s" % (str(e),),
+                errcode=Codes.FORBIDDEN,
+            )
+
+        try:
+            claims.validate(leeway=120)  # allows 2 min of clock skew
+
+            # Enforce the old behavior which is rolled out in productive
+            # servers: if the JWT contains an 'aud' claim but none is
+            # configured, the login attempt will fail
+            if claims.get("aud") is not None:
+                if self.jwt_audiences is None or len(self.jwt_audiences) == 0:
+                    raise InvalidClaimError("aud")
+        except JoseError as e:
+            raise LoginError(
+                403,
+                "JWT validation failed: %s" % (str(e),),
+                errcode=Codes.FORBIDDEN,
+            )
+
+        user = claims.get(self.jwt_subject_claim, None)
+        if user is None:
+            raise LoginError(403, "Invalid JWT", errcode=Codes.FORBIDDEN)
+
+        user_id = UserID(user, self.hs.hostname).to_string()
+
+        # If the account has been deactivated, do not proceed with the login
+        # flow.
+        try:
+            deactivated = await self._main_store.get_user_deactivated_status(user_id)
+        except StoreError:
+            # JWT lazily creates users, so they may not exist in the database yet.
+            deactivated = False
+        if deactivated:
+            raise UserDeactivatedError("This account has been deactivated")
```

Comment:
#12276 put this in the `_complete_login` flow of the rest servlet, which kind of seems better, but I'm a bit nervous that there's some flow there I don't understand.

Reply:
I think we *don't* do it there because of password providers, but I think this is confusing -- if you deactivate the user in Synapse it shouldn't matter if their password is still fine in the upstream password provider?

Reply:
IMO  all login methods should be call after specific validation also a global validation like [`validate_login`](https://github.com/matrix-org/synapse/blob/0602fc9f5265ed11534ef54a7fe904da8ecfdfb4/synapse/handlers/auth.py#L1128). There, global checks can be performed, such as whether a user is deactivated.

If a user is deactivated in synapse, It should not be able to log in. No matter if the user is active in the password provider or not. Otherwise I as the operator of Synapse have no more influence on the users and abuse or similar.

Reply:
> If a user is deactivated in synapse, It should not be able to log in. No matter if the user is active in the password provider or not. Otherwise I as the operator of Synapse have no more influence on the users and abuse or similar.

Just to be clear -- this is exactly what I said. If the user is deactivated in Synapse they'll be blocked from logging in regardless of what the password provider says.